File: recipes-to-json.lua
#!/usr/bin/env lua

-- convert factorio recipe.lua files to json

json = require("json")

data = {}

-- the factorio lua files call this method, so we have to implement it

function data.extend (target, new_data)
	-- append rows from new_data to target
	-- TODO: handle named table rows
	for i=1,#new_data do
        target[#target+1] = new_data[i]
    end
end

-- TODO: enumerate lua file list from directory
dofile("base/prototypes/recipe/ammo.lua")
dofile("base/prototypes/recipe/demo-furnace-recipe.lua")
dofile("base/prototypes/recipe/demo-turret.lua")
dofile("base/prototypes/recipe/fluid-recipe.lua")
dofile("base/prototypes/recipe/inserter.lua")
dofile("base/prototypes/recipe/recipe.lua")
dofile("base/prototypes/recipe/turret.lua")
dofile("base/prototypes/recipe/capsule.lua")
dofile("base/prototypes/recipe/demo-recipe.lua")
dofile("base/prototypes/recipe/equipment.lua")
dofile("base/prototypes/recipe/furnace-recipe.lua")
dofile("base/prototypes/recipe/module.lua")

-- TODO: possibly pretty-print the json, with indenting and newlines
print (json.encode( data ))

File: recipes-to-json.pl
#!/usr/bin/perl -p

# convert factorio recipe.lua files to json

# expects to be fed a .lua file that looks like this:
# data:extend(
# {
#   {
#     type = "recipe",
#     name = "player-port",
#     enabled = "false",
#     ingredients =
#     {
#       {"electronic-circuit", 10},
#       {"iron-gear-wheel", 5},
#       {"iron-plate", 1 }
#     },
#     result = "player-port"
#   },
#   {
#     type = "recipe",
#     name = "fast-transport-belt",
#     enabled = "false",
#     ingredients =
#     {
#       {"iron-gear-wheel", 5},
#       {"basic-transport-belt", 1}
#     },
#     result = "fast-transport-belt"
#   }
# }


# TODO: find a solution for trailing commas inside arrays

# TODO: handle whole file instead of line-by-line, for better bracket matching

s/^data:extend.$//g;
s/^\)$//g;

s/([^ ]*) =/"\1":/g;
s/(?<!^  ){/[/g;
s/(?<!^  )}/]/g;
